src/NHapi.Base/Parser/PipeParser.cs

```diff
@@ -24,885 +24,960 @@
+                    Convert.ToString(ec.FieldSeparator));
+                wholeFieldNine = fields[8];
+
+                // message structure is component 3 but we'll accept a composite of 1 and 2 if there is no component 3 ...
+                //      if component 1 is ACK, then the structure is ACK regardless of component 2
+                var comps = Split(wholeFieldNine, Convert.ToString(ec.ComponentSeparator));
+                if (comps.Length >= 3)
+                {
+                    messageStructure = comps[2];
+                }
+                else if (comps.Length > 0 && comps[0] != null && comps[0].Equals("ACK"))
+                {
+                    messageStructure = "ACK";
+                }
+                else if (comps.Length == 2)
+                {
+                    explicityDefined = false;
+                    messageStructure = comps[0] + "_" + comps[1];
+                }
+
+                /*else if (comps.length == 1 && comps[0] != null && comps[0].equals("ACK")) {
+                 messageStructure = "ACK"; //it's common for people to only populate component 1 in an ACK msg
+                 }*/
+                else
+                {
+                    var buf = new StringBuilder("Can't determine message structure from MSH-9: ");
+                    buf.Append(wholeFieldNine);
+                    if (comps.Length < 3)
+                    {
+                        buf.Append(" HINT: there are only ");
+                        buf.Append(comps.Length);
+                        buf.Append(" of 3 components present");
+                    }
+
+                    throw new HL7Exception(buf.ToString(), ErrorCode.UNSUPPORTED_MESSAGE_TYPE);
+                }
+            }
+            catch (IndexOutOfRangeException e)
+            {
+                throw new HL7Exception("Can't find message structure (MSH-9-3): " + e.Message, ErrorCode.UNSUPPORTED_MESSAGE_TYPE);
+            }
+
+            return new MessageStructure(messageStructure, explicityDefined);
+        }
+
+        /// <summary>
+        /// A struct for holding a message class string and a boolean indicating whether it
+        /// was defined explicitly.
+        /// </summary>
+        private class MessageStructure
+        {
+            public MessageStructure(string theMessageStructure, bool isExplicitlyDefined)
+            {
+                Structure = theMessageStructure;
+                ExplicitlyDefined = isExplicitlyDefined;
+            }
+
+            public string Structure { get; }
+
+            public bool ExplicitlyDefined { get; }
+        }
+
+        private class AnonymousClassPredicate : FilterIterator.IPredicate
+        {
+            private PipeParser enclosingInstance;
+
+            public AnonymousClassPredicate(PipeParser enclosingInstance)
+            {
+                InitBlock(enclosingInstance);
+            }
+
+            public PipeParser Enclosing_Instance

```

Comment:
do we need the field, can this be `{ get; private set; }` ?

Reply:
done